circuit/src/gadgets/main.rs

```diff
@@ -0,0 +1,950 @@
+		let fixed =
+			[F::one(), F::zero(), F::zero(), F::zero(), F::zero(), F::zero(), -F::one(), F::zero()];
+		let main_chip = MainChip::new(advices, fixed);
+		main_chip.synthesize(common, &config.selector, layouter)?;
+
+		Ok(())
+	}
+}
+
+/// Chip for is_equal operation
+pub struct IsEqualChipset<F: FieldExt> {
+	x: AssignedCell<F, F>,
+	y: AssignedCell<F, F>,
+}
+
+impl<F: FieldExt> IsEqualChipset<F> {
+	/// Create new IsEqualChipset
+	pub fn new(x: AssignedCell<F, F>, y: AssignedCell<F, F>) -> Self {
+		Self { x, y }
+	}
+}
+
+impl<F: FieldExt> Chipset<F> for IsEqualChipset<F> {
+	type Config = MainConfig;
+	type Output = AssignedCell<F, F>;
+
+	fn synthesize(
+		self, common: &crate::CommonConfig, config: &Self::Config, mut layouter: impl Layouter<F>,
+	) -> Result<Self::Output, Error> {
+		// We should satisfy the equation below
+		// x - y = 0
+
+		let diff = layouter.assign_region(
+			|| "assign_values",
+			|region| {
+				let diff = self.x.clone().value().cloned() - self.y.value();
+
+				let mut ctx = RegionCtx::new(region, 0);
+				let diff = ctx.assign_advice(common.advice[0], diff)?;
+				Ok(diff)
+			},
+		)?;
+
+		let is_zero_chip = IsZeroChipset::new(diff);
+		let res = is_zero_chip.synthesize(common, config, layouter)?;
+
+		Ok(res)
+	}
+}
+
+/// Chip for is_zero operation
+pub struct IsZeroChipset<F: FieldExt> {
+	x: AssignedCell<F, F>,
+}
+
+impl<F: FieldExt> IsZeroChipset<F> {
+	/// Create new IsZeroChipset
+	pub fn new(x: AssignedCell<F, F>) -> Self {
+		Self { x }
+	}
+}
+
+impl<F: FieldExt> Chipset<F> for IsZeroChipset<F> {
+	type Config = MainConfig;
+	type Output = AssignedCell<F, F>;
+
+	fn synthesize(
+		self, common: &crate::CommonConfig, config: &Self::Config, mut layouter: impl Layouter<F>,
+	) -> Result<Self::Output, Error> {
+		// We should satisfy the equation below
+		// 1 - x * x_inv = 1
+		// x * x_inv = 0
+		// x * x_inv + res - 1 = 0
+
+		// Witness layout:
+		// | A   | B     | C   | D   | E  |
+		// | --- | ----- | --- | --- | ---|
+		// | x   | x_inv | res |     |    |
+
+		let (zero, x_inv, res) = layouter.assign_region(
+			|| "assign_values",
+			|region| {
+				let x_inv = self.x.clone().value().map(|v| v.invert().unwrap_or(F::zero()));
+				let res = Value::known(F::one()) - self.x.clone().value().cloned() * x_inv.clone();
+
+				let mut ctx = RegionCtx::new(region, 0);
+				let zero = ctx.assign_advice(common.advice[0], Value::known(F::zero()))?;
+				let x_inv = ctx.assign_advice(common.advice[1], x_inv)?;
+				let res = ctx.assign_advice(common.advice[2], res)?;
+				Ok((zero, x_inv, res))
+			},
+		)?;
+
+		// [a, b, c, d, e]
+		let advices = [self.x.clone(), x_inv, res.clone(), zero.clone(), zero];
+		// [s_a, s_b, s_c, s_d, s_e, s_mul_ab, s_mul_cd, s_constant]
+		let fixed =
+			[F::zero(), F::zero(), F::one(), F::zero(), F::zero(), F::one(), F::zero(), -F::one()];
+		let main_chip = MainChip::new(advices, fixed);
+		main_chip.synthesize(common, &config.selector, layouter)?;
```

Comment:
Constraints here are incomplete. Consider the case where:
x = 1
x_inv = 0
res = 1
constrant `x * x_inv + res - 1` would still be satisfied.
We need an additional constraint to ensure that either `x` or `res` are zero

Reply:
Done in https://github.com/eigen-trust/protocol/pull/132/commits/f152771ee60f10afc1a27411a5e3e7debb0e1d87